### PR TITLE
Introduce the Results<TResult1, TResult2, TResultN> union types

### LINKED ISSUE #40672: Introduce way for route handler delegates (Minimal APIs) to return union results (i.e. multiple `IResult` types)
# Overview
Introduce the capability for route handler delegates to declare they return a union of multiple concrete implementations of `IResult` via their signature. This would provide compile-time checking that the delegate returns only the types declared, and when combined with #40646 allows the possible result types for the endpoint to be described in `ApiExplorer` automatically without the need for adding explicit metadata.

# Problem
Today, the type a route handler delegate declares it returns in its signature governs the default behavior of the framework when that delegate is invoked to handle a request. Returned types that implement `IResult` have their `Execute` method called once returned from the delegate, and all other types are serialized to the HTTP response body as JSON (including support for unwrapping `Task<T>` results>.

For very simple APIs this is satisfactory. Consider the following API:
``` c#
app.MapGet("/todos", async (TodoDb db) => await db.Todos.ToListAsync());
```
The passed route handler delegate will have its return type inferred by the compiler as `Task<List<Todo>>`. The framework will automatically JSON serialize the `List<Todo>` return value to the response body with a status code of **200 OK**.

Very often however, APIs need to be able to return different responses depending on the input, etc. Consider the following API:
``` c#
app.MapGet("/todos/{id}", async (int id, TodoDb db) =>
{
    await db.Todos.FindAsync(id) switch
    {
        Todo todo => Results.Ok(todo),
        _ => Results.NotFound()
    };
});
```
This API returns either a **200 OK** `IResult` if the `Todo` with the supplied id is found, or a **404 Not Found** `IResult` if no matching `Todo` is found. All return paths in the delegate must now return `IResult`, as the delegate can have only a single return type (which is inferred by the compiler in this case as both the `Results.Ok()` and `Results.NotFound()` methods return `IResult`). The framework observes that the delegate returns `IResult` and as such implements the behavior of calling the `Execute` method on whichever `IResult` instance the delegate returns when handling a request.

In some cases, the API will need to explicitly specify the return type of the delegate as the compiler will be unable to infer it. Consider the following API that uses a custom `IResult` type for one of the delegate's return paths. In order to enable returning different `IResult` types, the method must declare its return type as just `IResult`, meaning there's no information via the delegate signature as to exactly **which** `IResult` implementing types could be returned:
``` c#
app.MapGet("/todos/{id}", async IResult (int id, TodoDb db) =>
{
    await db.Todos.FindAsync(id) switch
    {
        Todo todo => Results.Ok(todo),
        _ => new MyCustomResult(404, "No Todo Found")
    };
});

class MyCustomResult : IResult
{
    public MyCustomResult(int statusCode, string message)
    {
        StatusCode = statusCode;
        Message = message;
    }
    public int StatusCode { get; } = 200;
    public string Message { get; } = "Not Found";
    public async Task ExecuteAsync(HttpContext httpContext)
    {
        httpContext.Response.Status = StatusCode;
        await httpContext.Response.Body.WriteAsync(Message);
    }
}
```

Without the concrete type information of the possible return types observable from the delegate signature, the framework is unable to implement any automatic behavior based on what the delegate might return, which is required along with #40646 to enable the API to be automatically more completely described in `ApiExplorer` (for the purposes of Swagger UI and OpenAPI documents).

# Proposed Solution

## Option 1 (Today?): Introduce an `IResult` union type in the framework, e.g. `Results<T1, T2, TN>`

> ℹ️ There is an example implementation of this option in the [`Results<T1, T2, ..>` type in the MinimalApis.Extension library](https://github.com/DamianEdwards/MinimalApis.Extensions/blob/main/src/MinimalApis.Extensions/Results/ResultsOfT.cs)

We would introduce a new generic type in the ASP.NET Core framework that implements `IResult` and accepts multiple generic type arguments (constrained to `IResult`) for the different result types a delegate can possibly return. This type would be implicitly convertible from any of it's generic argument types to itself, such that it need only be declared as the return type of delegate, and never actually created by user code.

**ResultsOfT.cs**
``` c#
public sealed class Results<TResult1, TResult2, TResult3, TResultN> : IResult
    where TResult1 : IResult
    where TResult2 : IResult
    where TResult3 : IResult
{
    private Results(IResult actualResult) => ActualResult = actualResult;

    public IResult ActualResult { get; }

    public async Task ExecuteAsync(HttpContext httpContext) => await ActualResult.ExecuteAsync(httpContext);

    public static implicit operator Results<TResult1, TResult2, TResult3>(TResult1 result) => new(result);
    public static implicit operator Results<TResult1, TResult2, TResult3>(TResult2 result) => new(result);
    public static implicit operator Results<TResult1, TResult2, TResult3>(TResult3 result) => new(result);
}
```

Consider the following API which has been updated from before to now declare that it returns either `OkResult<Todo>` or `NotFoundResult`:
``` c#
app.MapGet("/todos/{id}", async Task<Results<OkResult<Todo>, NotFoundResult>> (int id, TodoDb db) =>
{
    await db.Todos.FindAsync(id) switch
    {
        Todo todo => new OkResult<Todo>(todo),
        _ => new NotFoundResult()
    };
});
```
The compiler will ensure that only `IResult` types declared in the delegate's return signature are returned, and the `Results<T1, T2, TN>` type takes care of the implicit conversion of the declared concrete result types to itself. Additionally, when this delegate is observed by the framework, the full type information about which result types it can return is preserved, allowing for more capabilities regarding the default behavior employed.

A more complete example of an Minimal APIs application utilizing this approach can be found in [this sample of the MinimalApis.Extensions library](https://github.com/DamianEdwards/MinimalApis.Extensions/blob/main/samples/TodosApi.Dapper/TodosApi.cs).

## Option 2 (future?): Leverage a C# union type feature
If C# introduces support for declaring and utilizing [union types](https://github.com/dotnet/csharplang/issues/113), that could be utilized directly. It's similar to the approach in option 1 but a type would not be required in the framework.

The details of this would of course depend on how union types are implemented in C#, but consider the following example for a rough idea:
``` c#
// Compiler inferred union type (inferred anonymous union type)
app.MapGet("/todos/{id}", async union (int id, TodoDb db) =>
{
    await db.Todos.FindAsync(id) switch
    {
        Todo todo => new OkResult<Todo>(todo),
        _ => new NotFoundResult()
    };
});

// Union type declared inline (anonymous union type)
app.MapGet("/todos/{id}", async (OkResult<Todo> | NotFoundResult) (int id, TodoDb db) =>
{
    await db.Todos.FindAsync(id) switch
    {
        Todo todo => new OkResult<Todo>(todo),
        _ => new NotFoundResult()
    };
});

// Explicitly defined union type
app.MapGet("/todos/{id}", async Task<GetResult<Todo>> (int id, TodoDb db) =>
{
    await db.Todos.FindAsync(id) switch
    {
        Todo todo => new OkResult<Todo>(todo),
        _ => new NotFoundResult()
    };
});

enum struct GetResult<T> { OkResult<T> Ok, NotFoundResult NotFound }
```

### PR DESCRIPTION
**Pull request by DamianEdwards**
This PR introduces just the `Results<TResult1, TResult2, TResultN>` union types themselves. The types are code generated via a tool, along with their unit tests.

When #40646 lands we can update the in-box `IResult` implementations to describe themselves via emitted metdata, and update the `Results<TResult1, TResult2, TResultN>` union types to pass through that metadata from the `TResultN` type args, along with introducing the new `Results.Typed` factory methods for creating results and preserving the concrete types.

Until then, the union types allow all the possible `IResult` return types of a route handler delegate to be declared explicitly and thus checked at compile for type safety, e.g.:

``` c#
app.MapGet("/todos/{id}", async Task<Results<OkObjectHttpResult, NotFoundObjectHttpResult>> (int id, TodoDb db) =>
{
    return await db.Todos.FindAsync(id)
        is Todo todo
            ? (OkObjectHttpResult)Results.Ok(todo)
            : (NotFoundObjectHttpResult)Results.NotFound();
});
```

Fixes #40672

### CODE REVIEW

**Comment by brunolins16** [comment]
Is not possible (or make sense) use a T4 template instead of introduce a new project to generate the code?

File: src/Http/samples/MinimalSample/Program.cs [hunk 1]
```diff
@@ -19,6 +19,17 @@
 string SayHello(string name) => $"Hello, {name}!";
 app.MapGet("/hello/{name}", SayHello);
 
+app.MapGet("/null-result", () => (IResult)null);
```
Comment (davidfowl) [inline src/Http/samples/MinimalSample/Program.cs:22]:
```suggestion
app.MapGet("/null-result", IResult () => null);
```

**Comment by DamianEdwards** [comment]
> Is not possible (or make sense) use a T4 template instead of introduce a new project to generate the code?

I think to date the repo has used console apps for code generation in numerous places (Kestrel, Developer Exception Page, etc.) so I followed that pattern. I'm not super fussed other than it would mean more work to convert what I've done in this PR to use T4 instead, but there'll still be a mix in the repo.

File: src/Http/Http.Results/README.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+## Development Setup
+
+The `Results<TResult1, TResult2, TResultN>` union types are generated. Modify and run the [ResultsOfTGenerator](tools/ResultsOfTGenerator/) tool to generate an updated `ResultsOfT.cs` class file.
+
+Run the following command in `src\Http\Http.Results\tools\ResultsOfTGenerator`:
+
+```
+dotnet run
+```
```
Comment (halter73) [inline src/Http/Http.Results/README.md:13]: Can we also add a test that fails if the generator and generated code get out of sync like [Kestrel's GeneratedCodeIsUpToDate test](https://github.com/dotnet/aspnetcore/blob/a450cb69b5e4549f5515cdb057a68771f56cefd7/src/Servers/Kestrel/Kestrel/test/GeneratedCodeTests.cs#L21-L77)?

File: src/Http/Http.Results/tools/ResultsOfTGenerator/Program.cs [hunk 3]
```diff
@@ -0,0 +1,875 @@
...
+        writer.WriteIndentedLine("public async Task ExecuteAsync(HttpContext httpContext)");
+        writer.WriteIndentedLine("{");
+        writer.WriteIndentedLine(2, "ArgumentNullException.ThrowIfNull(httpContext, nameof(httpContext));");
+        writer.WriteLine();
+        writer.WriteIndentedLine(2, "if (Result is null)");
+        writer.WriteIndentedLine(2, "{");
+        writer.WriteIndentedLine(3, "throw new InvalidOperationException(\"The IResult assigned to the Result property must not be null.\");");
+        writer.WriteIndentedLine(2, "}");
+        writer.WriteLine();
+        writer.WriteIndentedLine(2, "await Result.ExecuteAsync(httpContext);");
```
Comment (halter73) [inline src/Http/Http.Results/tools/ResultsOfTGenerator/Program.cs:116]:
```suggestion
        writer.WriteIndentedLine("public Task ExecuteAsync(HttpContext httpContext)");
        writer.WriteIndentedLine("{");
        writer.WriteIndentedLine(2, "ArgumentNullException.ThrowIfNull(httpContext, nameof(httpContext));");
        writer.WriteLine();
        writer.WriteIndentedLine(2, "if (Result is null)");
        writer.WriteIndentedLine(2, "{");
        writer.WriteIndentedLine(3, "throw new InvalidOperationException(\"The IResult assigned to the Result property must not be null.\");");
        writer.WriteIndentedLine(2, "}");
        writer.WriteLine();
        writer.WriteIndentedLine(2, "return Result.ExecuteAsync(httpContext);");
```

It's a little bit more efficient if we skip the tail await at the small cost of losing the stack frame if there are any exceptions.

File: src/Http/Http.Results/README.md (hunk 2 above)
Comment (DamianEdwards) [reply src/Http/Http.Results/README.md:13]: Done!

File: src/Http/Http.Results/tools/ResultsOfTGenerator/Program.cs [hunk 4]
```diff
@@ -0,0 +1,909 @@
...
+        writer.WriteLine("namespace Microsoft.AspNetCore.Http;");
+        writer.WriteLine();
+
+        for (int i = 1; i <= typeArgCount; i++)
+        {
+            // Skip first as we don't have a Results<TResult1> class
+            if (i == 1)
+            {
+                continue;
+            }
```
Comment (halter73) [inline src/Http/Http.Results/tools/ResultsOfTGenerator/Program.cs:71]: Nit:

```suggestion
        // Skip 1 as we don't have a Results<TResult1> class
        for (int i = 2; i <= typeArgCount; i++)
        {
```

**Review by halter73 — APPROVED** [review]

**Pull request merged**